### PR TITLE
Test the library when malloc(0) returns NULL

### LINKED ISSUE #137: different behavior in `calloc` on different libraries when trying to allocate 0 bytes
### Description

[man calloc](https://linux.die.net/man/3/calloc) describes the following:
>If nmemb or size is 0, then calloc() returns either NULL, or a unique pointer value that can later be successfully passed to free().

This causes different behavior when calling the library wiht 0 bytes to allocate.
For example:
a call to  `psa_import_key()` will result in an error when `slen` is 0 because in `prepare_raw_data_slotoreoare_data_slot()` we will have:
~~~
    /* Allocate memory for the key */
    raw->bytes = PSA_BITS_TO_BYTES( bits );
    raw->data = mbedtls_calloc( 1, raw->bytes  );
    if( raw->data == NULL )
    {
        raw->bytes = 0;
        return( PSA_ERROR_INSUFFICIENT_MEMORY );
    }
~~~

and a call to `psa_key_derivation()` with the same empty key could result in a failure because in `psa_generator_tls12_prf_setup()`:
~~~
   tls12_prf->key = mbedtls_calloc( 1, key_len + 1 );
    if( tls12_prf->key == NULL )
        return( PSA_ERROR_INSUFFICIENT_MEMORY );
~~~

If using `memory_buffer_alloc` feature, the calloc calls would return `NULL` and not a valid pointer because:
~~~
    if( n == 0 || size == 0 || len / n != size )
        return( NULL );
~~~

There are several ways to fix this:
1. Don't call `mbedtls_calloc()` for a zero length -> This could cause crashes in the library for some cases where the data cannot be zero lengthed
2. Always add one more to the size to allocate -> This adds some additional RAM. One could argue this is insignificant.
3. Have the `memory_buffer_alloc()` return a 4 byte aligned pointer for zero length allocation request. This perhaps will fix the specific use case, but on other platforms, that have some calloc function that returns `NULL`, as it's allowed,the error will remain.
### Issue request type

    [ ] Question
    [ ] Enhancement
    [x] Bug

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
On our main test platforms (Linux, FreeBSD), malloc(0) returns a non-NULL pointer. With `MBEDTLS_MEMORY_BUFFER_ALLOC_C`, `calloc(0,_)` does return NULL, but memory_buffer_alloc doesn't play well with UBSan. Add a test component with UBSan where `calloc` returns NULL when asked to allocate 0 bytes, wrapping around the platform `calloc`.

Fix #137: now it's tested.

Should be backported to mbedtls LTS branches.

### CODE REVIEW

**Comment by Patater** [comment]
#257 is merged. Restarting CI.

File: programs/test/selftest.c [hunk 1]
```diff
@@ -77,6 +79,103 @@
...
+    }
+
+    if( verbose )
+        mbedtls_printf( "\n" );
+    mbedtls_free( empty1 );
+    mbedtls_free( empty2 );
+    mbedtls_free( buffer1 );
+    mbedtls_free( buffer2 );
+    mbedtls_free( overflow );
+    return( failures );
```
Comment (Patater) [inline programs/test/selftest.c]: What sorts of platforms would this fail on? Why are we checking exactly? What should we or our users do if we find a platform where this self-test fails? (Implement a calloc wrapper to make it work?)

> The main objective is to report whether calloc returns NULL for a size
of 0.

Yes, but so what? What is to be done with this knowledge?

File: tests/config-wrapper-malloc-0-null.h [hunk 2]
```diff
@@ -0,0 +1,39 @@
+/* config.h wrapper that forces calloc(0) to return NULL.
```
Comment (Patater) [inline tests/config-wrapper-malloc-0-null.h]: In commit message for "Add a test component with malloc(0) returning NULL".

> `Make this a separte job with UBSan (and ASan) to detect potential.`

s/separte/separate

and is there supposed to be a word after "potential" in this sentence?

File: tests/config-wrapper-malloc-0-null.h [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+ * Used for testing.
```
Comment (Patater) [inline tests/config-wrapper-malloc-0-null.h]: Why not put this file in `configs/` or, if that's not acceptable, `tests/configs/`? Having it at the `tests/` root is a bit messy, IMO.

File: tests/config-wrapper-malloc-0-null.h [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+ *  This file is part of mbed TLS (https://tls.mbed.org)
+ */
+
+#ifndef MBEDTLS_CONFIG_H
+/* Don't #define MBEDTLS_CONFIG_H, let config.h do it. */
+
+#include "mbedtls/config.h"
+
+#include <stdlib.h>
+static inline void *custom_calloc( size_t nmemb, size_t size )
```
Comment (Patater) [inline tests/config-wrapper-malloc-0-null.h]: This file implements a custom calloc. However, the commit message says "malloc". The all.sh test also says "malloc".

**Comment by gilles-peskine-arm** [comment]
CI reported failures. I'll fix these and also rebase on top of `development` now that the #257 has been merged.

File: programs/test/selftest.c (hunk 1 above)
Comment (gilles-peskine-arm) [reply programs/test/selftest.c]: It's a self-test. Self-tests don't check much that's useful. We don't expect the AES self-test to fail either — if it did, the unit tests would fail. The reason I added it here was to confirm that the test job that's supposed to test `calloc(0)=NULL` indeed has `calloc(0)=NULL`.

If the test fails, you know your `calloc` is broken. You know your platform better than we do, so we can't help you to fix it.

File: tests/config-wrapper-malloc-0-null.h (hunk 3 above)
Comment (gilles-peskine-arm) [reply tests/config-wrapper-malloc-0-null.h]: This is not a maintained configuration, or a meaningful configuration for users. So it should be under `tests`. I can move it to a subdirectory if you prefer but I don't think this makes `tests` any less messy.

File: tests/config-wrapper-malloc-0-null.h (hunk 4 above)
Comment (gilles-peskine-arm) [reply tests/config-wrapper-malloc-0-null.h]: `calloc` is in all but standardese a wrapper around `malloc`.

Comment (Patater) [reply tests/config-wrapper-malloc-0-null.h]: Looks a bit weird that it doesn't match names.

File: tests/config-wrapper-malloc-0-null.h (hunk 3 above)
Comment (Patater) [reply tests/config-wrapper-malloc-0-null.h]: Subdirectory would be good.

Scope creep: Do we have any other configs in the `configs` folder that are really just for testing and wouldn't really be meaningful configurations for users?

File: programs/test/selftest.c (hunk 1 above)
Comment (Patater) [reply programs/test/selftest.c]: What if we get `CALLOC(0): passed (distinct non-null)`? That is a valid and passing self-test result, isn't it? Does our other test fail to test what we want when that happens?

File: tests/config-wrapper-malloc-0-null.h (hunk 2 above)
Comment (gilles-peskine-arm) [reply tests/config-wrapper-malloc-0-null.h]: Fixed in rebase.

File: tests/config-wrapper-malloc-0-null.h (hunk 3 above)
Comment (gilles-peskine-arm) [reply tests/config-wrapper-malloc-0-null.h]: The other configurations are useful for users. I'll create a `test/configs`.

File: tests/config-wrapper-malloc-0-null.h (hunk 4 above)
Comment (gilles-peskine-arm) [reply tests/config-wrapper-malloc-0-null.h]: I changed the commit message, but kept some uses of malloc because “malloc” is a likely search term here.

File: programs/test/selftest.c (hunk 1 above)
Comment (gilles-peskine-arm) [reply programs/test/selftest.c]: `CALLOC(0): passed (distinct non-null)` is what you get on most if not all the platforms we test on today. I don't understand your question. The goal of this PR is to test on a platform with `CALLOC(0): passed (distinct non-null)`.

I didn't add a test run for `CALLOC(0): passed (same non-null)` because I don't think it matters to our code, whereas null vs non-null does.

Comment (Patater) [reply programs/test/selftest.c]: > The reason I added it here was to confirm that the test job that's supposed to test calloc(0)=NULL indeed has calloc(0)=NULL.

What if the platform doesn't have `calloc(0)=NULL`? Is the test job that's supposed to test `calloc(0)=NULL` invalid or shouldn't be run?

File: programs/test/selftest.c [hunk 5]
```diff
@@ -77,6 +79,105 @@
...
+{
+    int failures = 0;
+    void *empty1 = mbedtls_calloc( 0, 1 );
+    void *empty2 = mbedtls_calloc( 0, 1 );
+    void *buffer1 = mbedtls_calloc( 1, 1 );
+    void *buffer2 = mbedtls_calloc( 1, 1 );
+    /* Attempt to allocate a size that's larger than size_t. This
+     * should fail. With AddressSanitizer, this crashes the program by
+     * default; to avoid this, set the environment variable
+     * ASAN_OPTIONS=allocator_may_return_null=1 when running this program.
```
Comment (gilles-peskine-arm) [inline programs/test/selftest.c]: On the one hand, I don't like having the self-test program fail when run with ASan.

On the other hand, this is the one aspect of this self-test that might be plausibly broken on an embedded platform with a custom `calloc`, and it wouldn't cause functional bugs but could be a major security issue (e.g. convince the memory allocator that you want to allocate 4 bytes, but make mbedtls code think it's allocated SIZE_MAX+4 bytes).

File: programs/test/selftest.c (hunk 1 above)
Comment (gilles-peskine-arm) [reply programs/test/selftest.c]: I don't understand the question. The platform may or may not return NULL for malloc(0): both are valid. The objective of this PR is to make sure that we test the library on a platform where malloc(0) returns NULL. Since we only use platforms where malloc(0) is non-null, I simulate a platform where malloc(0) returns NULL by defining a wrapper function for `calloc` (we only use `calloc`, never `malloc`) that forcibly returns NULL when told to allocate 0 bytes. So now there is one component of `all.sh` that runs the unit tests with `calloc` returning NULL for 0 bytes. This component also runs the selftest and I confirm that I didn't mess up the way I set up the wrapper by visually inspecting that the selftest prints `NULL`. I didn't test that in `all.sh` because it's not worth the trouble to do it with our current tooling.

File: programs/test/selftest.c (hunk 5 above)
Comment (gilles-peskine-arm) [reply programs/test/selftest.c]: And in fact GCC's warning covers the case where the product of the parameters to `calloc` overflows. So if we perform this test, we need to do it without using values that the compiler can see. Since this is out of scope of this PR, I'll just remove this test.

File: tests/config-wrapper-malloc-0-null.h (hunk 3 above)
Comment (gilles-peskine-arm) [reply tests/config-wrapper-malloc-0-null.h]: Done in force-push.

**Comment by gilles-peskine-arm** [comment]
The remaining CI failures are unrelated to this PR (random DTLS failures, which is a known thing, and "export keys functionality" in `ssl-opt.sh` which is unrelated but I wouldn't have expected to fail here since the crypto CI uses the latest mbedtls).

File: programs/test/selftest.c (hunk 1 above)
Comment (Patater) [reply programs/test/selftest.c]: > make sure that we test the library on a platform where malloc(0) returns NULL.

This is checked only manually? or is there a test that will fail if the platform's malloc(0) doesn't return NULL?

Comment (gilles-peskine-arm) [reply programs/test/selftest.c]: Yes, I checked this manually. As a rule we don't have meta-tests that test that the tests are testing what we think they're testing, so I didn't do it here. It's like checking that a non-regression test fails once, and not keeping a CI run with a buggy product to check that it keeps catching the bug as the test code evolves.

File: tests/scripts/all.sh [hunk 6]
```diff
@@ -822,6 +822,22 @@ component_test_platform_calloc_macro () {
     make test
 }
 
+component_test_malloc_0_null () {
+    msg "build: malloc(0) returns NULL (ASan+UBSan build)"
+    scripts/config.pl full
+    scripts/config.pl unset MBEDTLS_MEMORY_BUFFER_ALLOC_C
```
Comment (AndrzejKurek) [inline tests/scripts/all.sh]: Since https://github.com/ARMmbed/mbedtls/pull/2469 you don't have to do that.

Comment (gilles-peskine-arm) [reply tests/scripts/all.sh]: I know, but that's only in mbedtls. The change isn't in mbed-crypto yet and this PR is for mbed-crypto.

Comment (AndrzejKurek) [reply tests/scripts/all.sh]: Right, sorry!

File: tests/scripts/all.sh [hunk 7]
```diff
@@ -822,6 +822,22 @@ component_test_platform_calloc_macro () {
...
+    make CC=gcc CFLAGS="'-DMBEDTLS_CONFIG_FILE=\"$PWD/tests/configs/config-wrapper-malloc-0-null.h\"' -O -Werror -Wall -Wextra -fsanitize=address,undefined" LDFLAGS='-fsanitize=address,undefined'
+
+    msg "test: malloc(0) returns NULL (ASan+UBSan build)"
+    make test
+
+    msg "selftest: malloc(0) returns NULL (ASan+UBSan build)"
+    # By default ASan terminates the process if called with a size that's
+    # insanely large. We do this deliberately in selftest, so disable this
+    # behavior.
```
Comment (Patater) [inline tests/scripts/all.sh]: We don't call calloc with a size that is extremely large anymore, so this comment needs updating.

File: programs/test/selftest.c (hunk 1 above)
Comment (Patater) [reply programs/test/selftest.c]: I wasn't requesting any meta tests. I was looking to ensure we weren't adding tests that had specific platform requirements and were running them on platforms that don't meet the requirements, whether self tests, test suites or otherwise.

I understand now what we are doing with the information from the self test, and that it is manual and not automatic. Thanks for explaining.

File: tests/scripts/all.sh (hunk 7 above)
Comment (gilles-peskine-arm) [reply tests/scripts/all.sh]: I amended the offending commit (which was the last one).

File: tests/scripts/all.sh [hunk 8]
```diff
@@ -822,6 +822,21 @@ component_test_platform_calloc_macro () {
...
+    scripts/config.pl full
+    scripts/config.pl unset MBEDTLS_MEMORY_BUFFER_ALLOC_C
+    make CC=gcc CFLAGS="'-DMBEDTLS_CONFIG_FILE=\"$PWD/tests/configs/config-wrapper-malloc-0-null.h\"' -O -Werror -Wall -Wextra -fsanitize=address,undefined" LDFLAGS='-fsanitize=address,undefined'
+
+    msg "test: malloc(0) returns NULL (ASan+UBSan build)"
+    make test
+
+    msg "selftest: malloc(0) returns NULL (ASan+UBSan build)"
+    # Just the calloc selftest. "make test" ran the others as part of the
+    # test suites.
```
Comment (Patater) [inline tests/scripts/all.sh]: How does "make test" avoid running the "calloc" selftest? I'm questioning if we need to call `make test`. I'm also questioning if we need to call calloc selftest separately if we decide we still want to run the whole `make test`

Comment (gilles-peskine-arm) [reply tests/scripts/all.sh]: Each test suite includes a call to the corresponding selftest function. This is why we don't bother to run `selftest` much in `all.sh`, just once to verify that the program works. This only misses the parts of `selftest.c` that aren't calling selftest functions from the various library modules.

We do need `make test` because the goal is to check that there isn't any null dereference (or similar undefined behavior) if `malloc(0)` returns null. This is very unlikely to come up in a crypto selftest. The places where it's likely to come up is in edge cases (empty input buffer, error recovery code).

Comment (Patater) [reply tests/scripts/all.sh]: OK, so `make test` is not calling the calloc self test, because we have no test suites that call the calloc self test.

**Comment by Patater** [comment]
CI failures are export keys, to be fixed by other PRs.
